libflashinfer/include/flashinfer/attention/generic/decode.cuh

```diff
@@ -650,111 +640,105 @@ gpuError_t SingleDecodeWithKVCacheDispatched(Params params, typename Params::DTy
   using DTypeQ = typename Params::DTypeQ;
   using DTypeKV = typename Params::DTypeKV;
   using DTypeO = typename Params::DTypeO;
+
   const uint32_t num_qo_heads = params.num_qo_heads;
   const uint32_t num_kv_heads = params.num_kv_heads;
   const uint32_t seq_len = params.kv_len;
 
-  constexpr uint32_t vec_size = std::max(16UL / sizeof(DTypeKV), HEAD_DIM / 32UL);
+  // AMD CDNA3 optimized vector size - prefer smaller vec_size for better occupancy
+  constexpr uint32_t vec_size = (HEAD_DIM < 256U)
+                                    ? std::max(8UL / sizeof(DTypeKV), HEAD_DIM / 64UL)
+                                    : std::max(8UL / sizeof(DTypeKV), HEAD_DIM / 32UL);
   constexpr uint32_t bdx = HEAD_DIM / vec_size;
+
   auto compute_capacity = GetCudaComputeCapability();
   static_assert(bdx <= 64U);
 
   DISPATCH_GQA_GROUP_SIZE(num_qo_heads / num_kv_heads, GROUP_SIZE, {
     constexpr uint32_t bdy = GROUP_SIZE;
-
-    // For AMD CDNA3, use fewer threads to reduce shared memory usage
     constexpr uint32_t num_threads =
         std::max(get_heuristic_num_threads(GROUP_SIZE, sizeof(DTypeKV)), bdx * bdy);
-
     constexpr uint32_t bdz = num_threads / (bdx * bdy);
 
-    // Use smaller tile size for AMD to reduce shared memory requirements
-    constexpr uint32_t tile_size_per_bdx = GROUP_SIZE == 1 ? (sizeof(DTypeKV) == 1 ? 2U : 4U) : 1U;
+    // AMD CDNA3 Reduce tile size to minimize shared memory usage
+    constexpr uint32_t tile_size_per_bdx = (GROUP_SIZE == 1U) ? 2U : 1U;
 
-    DISPATCH_COMPUTE_CAP_DECODE_NUM_STAGES_SMEM(compute_capacity, BASE_NUM_STAGES_SMEM, {
-      constexpr uint32_t NUM_STAGES_SMEM = std::min(BASE_NUM_STAGES_SMEM, 2U);
-      const uint64_t smem_size =
-          2U * NUM_STAGES_SMEM * bdy * tile_size_per_bdx * bdz * HEAD_DIM * sizeof(DTypeKV) +
-          2U * bdy * bdz * sizeof(float);
+    // This has been hard coded to 2U. Previous implementation involved a macro redirection that
+    // always resulted in 2U for H100 or CDNA3 architecture. Please take a look at
+    // gpu_iface/dispatch.cuh - DISPATCH_COMPUTE_CAP_DECODE_NUM_STAGES_SMEM macro
```

Comment:
Not really challenging why we should set `NUM_STAGES_SMEM` to `2`, but the heuristic for using `2` for CDNA3 is not clear to me here.